tree.c

```diff
@@ -1240,6 +1237,8 @@ void tree_sort_artists(void (*add_album_cb)(struct album *),
 			struct rb_node *t_node, *t_tmp;
 			struct album *album = to_album(l_node);
 
+			remove_album(album);
+			add_album(album);
```

Comment:
I'm not sure how much I like this, but it's okay, I think.

Reply:
Why is this required?

Reply:
The artists / albums / tracks are originally inserted into their respective list in sorted order based on whatever the current sorting rules are. This function is only used to force a resort when the "smart_artist_sort" and "sort_albums_by_name" settings are changed.

The remove function just unlinks the album from the artist's album tree, and the add function just links it back in sorted position. It doesn't involve album_new()/album_free() like you might have expected from the name.

As for removing/inserting from inside the `rb_for_each_safe` loop -- although some elements may be visited twice if they are moved forwards in the rb_tree, all elements will be processed at least once and at most twice, thanks to the `_safe` part.